e2store/src/era2.rs

```diff
@@ -139,8 +127,51 @@ impl Era2 {
     }
 }
 
+impl Iterator for Era2 {
+    type Item = AccountOrStorageEntry;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.pending_storage_entries > 0 {
+            self.pending_storage_entries -= 1;
+
+            let raw_storage_entry = match self.e2store_stream.next_entry() {
+                Ok(raw_storage_entry) => raw_storage_entry,
+                Err(err) => panic!("Failed to read next storage entry: {:?}", err),
+            };
+
+            let storage_entry = match StorageEntry::try_from(&raw_storage_entry) {
+                Ok(storage_entry) => storage_entry,
+                Err(err) => panic!("Failed to decode next storage entry: {:?}", err),
+            };
+            return Some(AccountOrStorageEntry::Storage(storage_entry));
+        }
+
+        let raw_account_entry = match self.e2store_stream.next_entry() {
+            Ok(raw_account_entry) => raw_account_entry,
+            Err(err) => match err {
+                // If we read to the end of the error file we should get this
+                err if underlying_io_error_kind(&err).is_some()
```

Comment:
it seems that this error handling should be done at the `E2StoreStream.next_entry()` level, in which case it should return `Result<Option<Entry>>` (or `Option<Result<Entry>>`, either would be fine there).